rainforest/tests.go

```diff
@@ -128,6 +130,56 @@ type testElementDetails struct {
 	Elements []testElement `json:"elements,omitempty"` // for embedded tests
 }
 
+type Wisp struct {
+	Version string `json:"version"`
+	Verbs   []Verb `json:"verbs"`
+}
```

Comment:
How do we handle different versions of Wisp in the future?

Reply:
Have you thought about how and when we update Wisp to have different verbs (and so the Wisp schema changes), how will the CLI support having different versions (either stored locally, or downloaded from the product)?

It is necessary to support prior versions so we can update the CLI+Wisp without everything breaking whenever we update.

Reply:
Currently the only thing that would need updating is the wisp structures.
If we add new fields it would just ignore them, so the failure would come from trying to reupload a downloaded test.
To fix this we would need to add any new field (or any old removed field from prior versions if any)as an optional field in the wisp structs.

Alternatively if we don't want to do that, we could accept the entire wisp as a json blob and store/reupload that.
This would require us then to delegate any and all validation to the backend as writing and maintaining a validator defeats the purpose of removing the structs.

I guess it depends on how involved we want the CLI to be in the contents of the wisp. Alternatively we could also expose an endpoint(somewhere) to validate wisp, which could allow us to accept a json blob on the client side and still be able to validate it when necessary.

While writing this, I realized that I was working on the assumption that client side validation will be required at some point(currently we're not really doing it, but the structures are a first step), so if that assumption is wrong, the json blob might be the right choice.

Reply:
@jbarber I tried changing the structures to a json blob as per our discussion, but it doesn't just work. The blocker here is that actions have an id which blows up backend when trying to update. Backend checks if an id is present and if it is it tries to find an element matching the id. Because some elements might be deleted, it can't find them and raises an error.

This worked with structures as I was ignoring the id, so it would never be send to backend and backend would create a new element. We discussed before that we might want to change backend to not blow up if there is a provided id, but the element is deleted.

My suggestion would be to either ship as is for now and revisit later, or update backend to not blow up when an id is provided for an element, but the element is deleted.
Updating backend would also have the added benefit of reusing existing elements when possible instead of always creating a new one.

After a quick check of the code, it seems like it wouldn't take much work to make this change backend.